File: symmetric_key_cryptography/js/playfair_cipher.js
// Before reading the playFairCipher code you should make sure you have a good
// understanding of the helper functions involved with the cipher, they are 
// pretty basic but important to understanding the 
function playFairCipher (keyword, message) {
    const alpha = 'abcdefghiklmnopqrstuvwxyz';
    const matrix = matrixCipher(keyword, alpha);
    const pairedChars = splitMessage(message);
    const m = matrix.matrix;
    const pos = matrix.positions

    const encryptedMessage = [];

    for (let i = 0; i < pairedChars.length; i++) {
        const pair = pairedChars[i];
        // Memoizing the position of the characters in order to stay DRY
        const pos0 = pos[pair[0]];
        const pos1 = pos[pair[1]]; 

        let newPos0;
        let newPos1;
        
        if (pos0[1] === pos1[1]) {
            // This is if the letters are both in the same column
            // Doing basic modular arithmatic to make sure that if the letters 
            // are at the end of the matrix that theyll come back to the beginning
            newPos0 = [(pos0[0] + 1) % m.length, pos0[1]];
            newPos1 = [(pos1[0] + 1) % m.length, pos1[1]];
        } else if (pos0[0] === pos1[0]) {
            // Like the last conditional this checks if the letters are in the
            // same row in the matrix
            newPos0 = [pos0[0], (pos0[1] + 1) % m[0].length];
            newPos1 = [pos1[0], (pos1[1] + 1) % m[0].length];
        } else if (pos0[1] !== pos1[1]) {
            // If the position of the two letters arent in the same row or column
            // in the matrix all you have to do is swap the column numbers and you 
            // have found the cipher letter that will be replacing the cur letter 
            newPos0 = [pos0[0], pos1[1]];
            newPos1 = [pos1[0], pos0[1]];
        } 
        const newPair = m[newPos0[0]][newPos0[1]].concat(m[newPos1[0]][newPos1[1]])
        encryptedMessage.push(newPair)
    }
    return encryptedMessage.join('');
}

function splitMessage(message) {
    const joined = message.split(' ').join('').toUpperCase();
    const splitMessage = [];

    let i = 0;
    while (i < joined.length) {

        // If both letters are the same the playfair cipher has you return
        // the first letter with an X following it.
        if (joined[i] === joined[i + 1]) {
            splitMessage.push(joined[i] + 'X');
            i += 1;
            continue;
        } 

        // Check if we are dealing with the last index in a string
        // if we are then attach it to an X. Ever letter needs a pair for the 
        // cipher to work.
        if(joined[i + 1] === undefined) {
            splitMessage.push(joined[i] + 'X');
            break;
        }

        splitMessage.push(joined[i] + (joined[i+ 1]))
        
        i += 2;
    }

    return splitMessage;
}

// Test to make sure you are developing the right return value in your splitMessage
// You want only pairs of 2, no undefineds, and not single letters on their own

// const message = 'secret messages';

// const messageResult = splitMessage(message);
// console.log(messageResult)


// This helper function will take in the keyword which is our ciphery key,
// And the alphabet which we will use to create a 5x5 matrix
// Dont forget to omit the letter J

function matrixCipher(keyword, alpha) {
    const lettersObj = {};
    const letters = keyword.concat(alpha).toUpperCase().split('');

    // this allows us to return multiple values;
    this.matrix = [];
    this.positions = {};

    alpha.split('').forEach((char) => {
        lettersObj[char] = false;
    });

    // Tracks the index of the next letter in the array
    let x = 0;
    for(let i = 0; i < 5; i++) {
        const matrixLayer = [];

        for(let j = 0; j < 5; j++) {
            const cur = letters[x];

            // Check hash if the letter has been used
            if(lettersObj[cur] !== true) {
                // Push the letter in the right position
                matrixLayer.push(letters[x]);
                lettersObj[cur] = true;

                // Save the position of the letter in the matrix for O(1) lookup
                // Instead of having to constantly loop through the matrix
                this.positions[letters[x]] = [i, j];
            } else {
                // If the letter has been used lower the count so we go to the 
                // next letter without losing an iteration
                j--
            }
            x = x + 1;
        }
        this.matrix.push(matrixLayer);
    };
    return this;
}

// Run the following code below to check for yourself, but there should be
// no repeats in the matrix, and the first letters should always start with 
// the cipher key. Also J should not be in there.

// const alpha = 'abcdefghiklmnopqrstuvwxyz';
// const keyword = 'keyword';

// const matrixResult = matrixCipher(keyword, alpha);

// console.log(matrixResult.matrix)
// console.log(matrixResult.positions)
//   [ ['k', 'e', 'y', 'w', 'o'],
//     ['r', 'd', 'a', 'b', 'c'],
//     ['f', 'g', 'h', 'i', 'l'],
//     ['m', 'n', 'p', 'q', 's'],
//     ['t', 'u', 'v', 'x', 'z'] ]


const message = 'secret message';
const keyword = 'keyword';

const cipher = playFairCipher(keyword, message);
console.log(cipher);

File: symmetric_key_cryptography/python/playfair_cipher.py
class Playfair:

    def __init__(self, keyword: str, message: str):
        super().__init__()
        self._keyword = keyword
        self._message = message
        self.alpha = 'abcdefghiklmnopqrstuvwxyz'
        self._matrix = []
        self._positions = {}
    
    def encrypt(self):
        # Populate the matrix
        self._matrix_cipher(self._keyword, self.alpha)
        # Turn the secret message into pairs of characters. 
        # Read the _split_message function to see rules regarding play_fairs pairs
        paired_chars = self._split_message(self._message)

        encrypted_message = []

        for i in range(len(paired_chars)):
            pair = paired_chars[i]
            # Memoize the position of the chracters in order to stay DRY
            pos_0 = self._positions[pair[0]]
            pos_1 = self._positions[pair[1]]

            if (pos_0[1] == pos_1[1]):
                # This is if the letters are both in the same column
                # Doing basic modular arithmatic to make sure that if the letters
                # are at the end of the matrix that they'll come back to the begining
                new_pos_0 = [(pos_0[0] + 1) % len(self._matrix), pos_0[1]]
                new_pos_1 = [(pos_1[0] + 1) % len(self._matrix), pos_1[1]]
            elif (pos_0[0] == pos_1[0]):
                # Like the previous conditional this checks if the letters are in
                # in the same row in the matrix
                new_pos_0 = [pos_0[0], (pos_0[1] + 1) % len(self._matrix[0])]
                new_pos_1 = [pos_1[0], (pos_1[1] + 1) % len(self._matrix[0])]
            elif (pos_0[1] != pos_1[1]):
                # If the position of the two letters arent in the same row or column
                # in the matrix all you have to do is swap the column numbers and you
                # have found the cipher letter that will be replacing the cur letter
                new_pos_0 = [pos_0[0], pos_1[1]]
                new_pos_1 = [pos_1[0], pos_0[1]]

            
            letter_1 = self._matrix[new_pos_0[0]][new_pos_0[1]]
            letter_2 = self._matrix[new_pos_1[0]][new_pos_1[1]]

            new_pair = letter_1 + letter_2

            encrypted_message.append(new_pair)

        return ''.join(encrypted_message)

    def _split_message(self, message: str):
        message_list = message.split(' ')
        joined = ''.join(message_list).upper()
        split_message = []

        i = 0
        while (i < len(joined)):
            if (joined[i] == joined[i + 1]):
                split_message.append(joined[i] + 'X')
                i += 1
                continue
            if (joined[i + 1] == None):
                split_message.append(joined[i] + 'X')
                break
        
            split_message.append(joined[i] + (joined[i + 1]))

            i += 2
        return split_message

    def _matrix_cipher(self, keyword: str, alpha: str):
        letters_obj = {}
        letters = list((keyword + alpha).upper())

        # Set all the letters to false in the hash
        for char in list(letters):
            letters_obj[char] = False
        
        # Tracks the index of the next letter in the array
        x = 0
        for i in range(5):
            matrix_layer = []

            # We use a while loop and the counter so that we can count backwards 
            # and forward depending on the matrix layer setup
            count = 0
            while(count < 5):
                cur = letters[x]
                # Check the hash if the letter has been used
                if letters_obj[cur] == False:
                    # print('hit it heres the count', count)

                    # Push the letter in the right position
                    matrix_layer.append(letters[x])
                    letters_obj[cur] = True 
                    # Save the position of the letter in the mattrix for O(1) lookup
                    # Instead of having to constantly loop through the matrix
                    self._positions[letters[x]] = [i, count]
                else:
                    count -= 1

                count += 1
                x += 1

            self._matrix.append(matrix_layer)

message = 'secret message';
key = 'keyword'
alpha = 'abcdefghijklmnopqrstuvwxyz'

encryption = Playfair(key, message)
print(encryption.encrypt())
